### ISSUE TITLE
 matplotlib 3 pyplot on MacOS bounces rocket icon in dock

### ISSUE THREAD

**Issue by defjaf** [matplotlib/matplotlib#12188]
### Bug report

**Bug summary**

Rocket icon appears in dock (and bounces) when `matplotlib.pyplot` is imported after installing matplotlib 3.0. In previous versions, this didn't happen (and still doesn't with `python2` on the same machine)

**Code for reproduction**

```python
import matplotlib.pyplot 
```

**Actual outcome**
Rocket icon appears in MacOS dock and bounces for a while.

**Expected outcome**
Nothing (no dock icon for previous versions).

**Matplotlib version**
  * Operating system: MacOS
  * Matplotlib version: 3
  * Matplotlib backend: any (seems to happen no matter what is set in matplotlibrc)
  * Python version: 3.7 (Python.org framework build)

Installed from pip.

I haven't seen any other complaints about this, so I suspect that there is something specific to my setup.

Perhaps related to [this ipython issue](https://github.com/ipython/ipython/issues/10137) (but this happens even from the plain python3 REPL.) I

**Comment by tacaswell** [comment]
Is python hung while it is bouncing?  If so I suspect that this is related to the font search issue (see https://github.com/matplotlib/matplotlib/issues/12176 / #12177 ).  We have a fix in, and there will likely be a 3.0.1 in a few weeks.   For now I suggest either going to an empty directory, doing the import there to let the font cache build.

**Comment by defjaf** [comment]
No, I don't think it's that. Python is running just fine with the icon in the dock.

(Aside: I do think I've also encountered that issue, however -- when diagnosing this, I tried to remove my `.matplotlib` directory and definitely found several issues with the system trying to set up fonts: 1. it stops whenever it finds a bad symlink anywhere in my filesystem; 2. Even after I fixed all of those, it still eventually failed, after more than 20 minutes, with `FileNotFoundError: [Errno 2] No such file or directory: '/Users/jaffe/.matplotlib/fontlist-v300.json'`)

**Comment by defjaf** [comment]
Perhaps I should ask if anyone is or isn't seeing this behaviour?

**Comment by defjaf** [comment]
See also [this post from Dr Drang](http://leancrew.com/all-this/2014/01/stopping-the-python-rocketship-icon/) regarding Python 2.7.

Note that the plist hack still works, but `matplotlib.use("Agg")` doesn't anymore. Could this be due to the fallback backend search?

But this brings up another question: I *don't* see this behaviour under Python 2.7 (and I don't remember seeing it at the time that the Dr Drang was written, around 2014). Is this setup-dependent? Or was there some other fix to 2.7?

**Comment by jklymak** [comment]
This is somewhat confusing as Matplotlib 3.0.0 does not run under python 2.7.  So did you mean Matplotlib 2.2.x doesn't bounce?  

I have a Mac and I never see the rocket icon after I set my backend to Qt5Agg.  Have you tried setting this in `~/.matplotlib/matplotlibrc`?  Its possible the new renderer fallback is not working properly...

This'll be hard for me to debug - I use conda for my install, not python.org + pip.  If you go that route, its entirely possible there are mismatched libraries, and you should consider upgrading all the non-python dependencies...  

Sorry to not be more helpful.

**Comment by defjaf** [comment]
To be clear:

- **Python 3 + Matplotlib 3**: bounces no matter what backend
- **Python 2.7 + Matplotlib 2.2x**: gives no icon if launched without a gui backend in `matplotlibrc`, bouncing icon if launched with gui backend
- **Python 3 + Matplotlib 2.2x**: I don’t have this setup on my machine right now, but I recall that  it behaves as on Python 2.7.

**Comment by jklymak** [comment]
> - Python 3 + Matplotlib 3: bounces no matter what backend

That is hard to understand, and my guess is that its always choosing the same backend, depsite your intentions

Can you put 

```python
import matplotlib
print(matplotlib.get_backend())
```
at the top of your script and check that its really giving the intended backend?

**Comment by defjaf** [comment]
Yep:

```
$ python3
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 26 2018, 23:26:24) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib
>>> matplotlib.get_backend()
'agg'
>>> import matplotlib.pyplot as plt
```

bounce - bounce - bounce

**Comment by gsportelli** [comment]
I can confirm that with Python 3 and Matplotlib 2.2.2 the icon does not show up.

**Comment by jklymak** [comment]
@defjaf   Sorry, I can't reproduce conda py3.7, mpl3.0.0.  I don't get any icons in the doc until I do plt.show().  Must be something funky w. the framework build.

**Comment by defjaf** [comment]
@jklymak So does that mean "wontfix" or do we need to dig further?

**Comment by jklymak** [comment]
Oh no absolutely not.  It means I probably won’t fix, but hopefully someone on the dev team uses non-conda on a Mac. 

Do you ever get the prompt back?   Can you check backend after you get the prompt back?

**Comment by defjaf** [comment]
@jklymak Aha, I think there's a misunderstanding: the bouncing icon does *not* hang the interpreter. The prompt returns immediately and it runs just fine.

**Comment by jklymak** [comment]
@tacaswell @anntzer is it possible the new backend code is importing and checking a backend before accepting the OP's choice of "Agg"?

**Comment by tacaswell** [comment]
if you do `get_backend()` again after importing pyplot what do you get?

If you start a fresh interpreter can you do:

```python
import matplotlib
dict.__getitem(matplotlib.rcParams, 'backend')
```

(which side steps the auto-fallback logic)

To check, `backend: 'agg'` is in your rcparams file?  It make be better to test with `import matplotlib; matplotlib.use('BACKEND_OF_CHOICE')` just to be sure!

One possible concern is that the osx backend is importing _just_ enough to trigger the rocket, but not enough to be selected?

It may also help to put some print statements in the `matplotlib.switch_backend` function to see when that is being called with what inputs.

@defjaf Thanks for sticking with us to sort this out!

**Comment by defjaf** [comment]
`get_backend` and the `matplotlib.rcParams` dict behave as expected (note that the syntax is somewhat different than you suggested):

```
$ python3
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 26 2018, 23:26:24) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib; matplotlib.rcParams.__getitem__('backend')
'agg'
>>> matplotlib.use('PDF'); matplotlib.rcParams.__getitem__('backend')
'pdf'
>>> import matplotlib.pyplot
```
bounce - bounce - bounce ...

**Comment by jklymak** [comment]
OK, my appologies - I can reproduce on my anaconda set up.  I don't get a rocket, but I get an app icon that flashes for a while and a beachball if I don't change focus back to the terminal.  Qt5Agg is the backend both before and after `import matplotlib.pyplot as plt`.   

This is definitely to do with the Framework build - if I use the non-framework build, and use Qt5Agg, the extra dock icon never shows up.  

The dock icon controls the visibility of the window that shows up if I call `plt.show()`.  So, somehow importing `pyplot` is now queing up the figure manager (?) in the Framework Build but not in the non-framework build.  Note again that Qt5Agg is my default backend. 

Reverting to 2.2.3 this doesn't happen.

**Comment by tacaswell** [comment]
er, I meant ` dict.__getitem__(matplotlib.rcParams, 'backend')`  The goal is to side-step https://github.com/matplotlib/matplotlib/blob/09e9504ed4e057318e2675475d23847900078eed/lib/matplotlib/__init__.py#L876-L880

The bouncing does not start until you import `pyplot` though?

I suspect the issues are in https://github.com/matplotlib/matplotlib/blob/09e9504ed4e057318e2675475d23847900078eed/lib/matplotlib/pyplot.py#L2331-L2339

**Comment by jklymak** [comment]
Just commenting all that out doesn't change the behaviour for me (and the if conditional is `None`)

**Comment by defjaf** [comment]
> er, I meant `dict.__getitem__(matplotlib.rcParams, 'backend')` 

Same (correct) behaviour as above.

> The bouncing does not start until you import `pyplot` though?

Exactly.

> I suspect the issues are in
> 
> [matplotlib/lib/matplotlib/pyplot.py](https://github.com/matplotlib/matplotlib/blob/09e9504ed4e057318e2675475d23847900078eed/lib/matplotlib/pyplot.py#L2331-L2339)
> Lines 2331 to 2339 in [09e9504](/matplotlib/matplotlib/commit/09e9504ed4e057318e2675475d23847900078eed)
>  # If rcParams['backend_fallback'] is true, and an interactive backend is 
>  # requested, ignore rcParams['backend'] and force selection of a backend that 
>  # is compatible with the current running interactive framework. 
>  if (rcParams["backend_fallback"] 
>          and dict.__getitem__(rcParams, "backend") in _interactive_bk 
>          and _get_running_interactive_framework()): 
>      dict.__setitem__(rcParams, "backend", rcsetup._auto_backend_sentinel) 
>  # Set up the backend. 
>  switch_backend(rcParams["backend"])

OK, I've played with this code a bit. 

The `if (rcParams["backend_fallback"]` block doesn't matter, but if I comment out `switch_backend(rcParams["backend"])` I *don't* get a bouncing with a non-GUI backend. However, if I do `matplotlib.use("MacOSX")` and then `import matplotlib.pyplot as plt`, I get the bouncing icon.

But note that this behaviour is still a bug, I think -- under 2.7, the icon didn't *bounce*.

**Comment by defjaf** [comment]
> Just commenting all that out doesn't change the behaviour for me (and the if conditional is `None`)

Hmmm: As per above, the `if`... block has no effect for me, but `switch_backend` does help, at least for non-GUI backends.

**Comment by jklymak** [comment]
Not surprisingly bisects to 56acb0f691f344

```
git bisect bad
56acb0f691f3447f7f15b257b164e0bc431760eb is the first bad commit
```

**Comment by jklymak** [comment]
The culprit looks to be:

https://github.com/matplotlib/matplotlib/blob/09e9504ed4e057318e2675475d23847900078eed/lib/matplotlib/pyplot.py#L214-L215

If I run this line I get the bouncing dock - so I guess `_get_running_interactive_framework` checks for the right frame work, even if MacOSX is not the framework...

**Comment by jklymak** [comment]
#12207 removes the check that opens the framework if using a non-GUI backend.  If you use a GUI backend, I guess this icon coming up upon import is expected behaviour (?)...

**Comment by defjaf** [comment]
Digging down further, I get the bouncing icon with `from matplotlib.backends import _macosx`

Apologies if I misunderstand, but will your proposed fix still give a *bouncing* icon when the GUI is started? Is that still a bug?

**Comment by jklymak** [comment]
My proposed fix will still open the icon when `_get_running_interactive_framework` is called as far as I can tell; @anntzer or @tacaswell can clarify, but I think it moves any framework incompatibility checks earlier in the process before the first draw.  So that may not be a bug if your backend requires the GUI.

**Comment by jklymak** [comment]
Hmm, with qt5Agg as my backend running `_get_running_interactive_framework()` returns `None` and `required_framework` is `qt5`.  So somehow the checks in `_get_running_interactive_framework():` aren't catching that I have `qt5` installed.   Maybe thats OK, because the function eventually returns "None", but not before it tries 
`from matplotlib.backends import _macosx` which is what pops up the window.  

Questions:  

1.  Should I be getting "qt5" out of this function, or is that just if someone is running matplotlib from a GUI that they have made? 
2. Should we be trying the matplotlib import w/o macosx running?  
3. Returning none, when there is a required backend never fails.  Is that OK?

**Comment by tacaswell** [comment]
It seems there are 2 bugs here:
 
 1. we are hitting something that is triggering the icon even for non OSX backend
 2. the icon bounces forever

I am concerned that we are just not understanding the OSX gui framework well enough and are starting it by asking if it is started, but in a way that leave it in a "I am turning me self on, please wait" state.

**Comment by defjaf** [comment]
> My proposed fix will still open the icon when `_get_running_interactive_framework` is called as far as I can tell; @anntzer or @tacaswell can clarify, but I think it moves any framework incompatibility checks earlier in the process before the first draw. So that may not be a bug if your backend requires the GUI.

To clarify: the **bouncing** is a remaining bug -- the icon shows up but doesn't _bounce_ under 2.7/2.2.x

**Comment by defjaf** [comment]
To reiterate, #12207 *doesn’t* fix the bouncing (which is an unwelcome change from 2.2.x).

**Comment by defjaf** [comment]
And to be yet more precise, the minimal working (failing) example for the bug appears to be:

    from matplotlib.backends import _macosx

* Under 2.2.x: icon in dock, *no bouncing*
* Under 3.0.0: icon in dock, *bouncing*

**Comment by jklymak** [comment]
#12207 fixed popping up the icon for a backend that didn’t need it.  I have no idea why if you do need a backend that will pop up the icon the bouncing happens, but I’m also  not clear why the icon is a problem. 

Note this is really only a symptom of the framework build on Mac OS.  A possible workaround is to not use the framework build and avoid the macOSx backend.

**Comment by tacaswell** [comment]
@jklymak I assume because it is _super_ annoying

@defjaf which of those are framework / non-framework python?

**Comment by jklymak** [comment]
It stops bouncing as soon as something is drawn. I assume what is happening is that import _macosx pops up the icon and it keeps bouncing until a plot window is created. I suspect that all that can be done is remove the macOSx framework check until draw time or somehow undo whatever application spinup import _macosx does during the check and reimport later.

**Comment by defjaf** [comment]
> @jklymak I assume because it is _super_ annoying

Indeed, and because it's a change from 2.2.x

> @defjaf which of those are framework / non-framework python?

They are all (I think) Framework, but Apple supplies 2.7.10 with macOS 10.13 (High Sierra; I assume that's about to change as of the imminent release of 10.14 (Mojave); 2.7.15 and 3.7 are python.org framework builds.

**Comment by dstansby** [comment]
Some more debugging: commenting out these lines stops the icon bouncing:

https://github.com/matplotlib/matplotlib/blob/09e9504ed4e057318e2675475d23847900078eed/src/_macosx.m#L2647-L2648

so there is something going on in `verify_framework` that is spawning the icon.

**Comment by terrah27** [comment]
> Perhaps I should ask if anyone else on MacOS is (or isn't) seeing this behaviour?

Yes! I just set up a new MacBook Pro and for every open jupyter notebook with matplotlib imported I get the bouncing spaceship.  I don't have this problem on my old MacBook at all or on my Ubuntu Desktop.

**Comment by tacaswell** [comment]
@terrah27 what are the OSX versions and matplotlib versions are on the two machines? 

We are pretty sure that this is related to interactions with the OSX window manager so it should not affect anyone on linux.

**Comment by terrah27** [comment]
I'm running macOS High Sierra 10.13.6 matplotlib 3.0.0 on the macbook with the issue.  I don't have the older machine with me to check the versions right now.

**Comment by jklymak** [comment]
How did you install matplotlib and jupyter?

**Comment by terrah27** [comment]
I used anaconda

**Comment by mrivers99** [comment]
Same issue here running macOS Mojave 10.14 and Matplotlib 3.0.0.

**Comment by louis925** [comment]
There is another issue on Spyder relates to this bounding rocket icon. When the rocket icon is flashing, the CPU is high. Any input to the console of Spyder will freeze until you click on the bouncing python rocket icon in dock. The Spyder version is 3.3.1. Python version is 3.6.6. The issue happens in Matplotlib 3.0.0 but not in 2.2.3.

**Comment by bakert** [comment]
I have a variation on this issue using matplotlib 3.0.0 on OSX 10.13.6 with python 3.6 or 3.7 (installed from macports).

When I start my flask webserver I get not one but two rocketship icons. The following code is implicated:

```
import matplotlib as mpl
# This has to happen before pyplot is imported to avoid needing an X server to draw the graphs.
# pylint: disable=wrong-import-position
mpl.use('Agg')
import matplotlib.pyplot as plt
```

They do bounce for a long time but not forever. Because Flask reloads automatically on detecting a change on the filesystem I get the beachball and a lot of bouncing every time I save a file and have to return focus to my text editor.

**Comment by bakert** [comment]
```
[gaghalfrunt ~] python
Python 3.7.0 (default, Jun 28 2018, 05:55:06) 
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib as mpl                                                    
>>> mpl.use('Agg')
>>> import matplotlib.pyplot as plt
```

gives the exact same behavior. The bouncing rocketship comes when I enter the final line.

**Comment by tacaswell** [comment]
I think there are actually two issues here,

1. why the rocket is bouncing
2. why if the user has used `mpl.use` we are going through auto-discovery (see https://github.com/matplotlib/matplotlib/issues/12188#issuecomment-428820116)

**Comment by laurentperrinet** [comment]
I can confirm the same behavior on

Python : 3.6.5 64bit [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
OS : Darwin 18.0.0 x86_64 i386 64bit
matplotlib : 3.0.0

I had written a question on [stackoverflow](https://stackoverflow.com/questions/52807478/matplotlib-pops-up-a-window-when-launched-from-jupyter) which is well described by this issue.

**Comment by jklymak** [comment]
OK, to clarify a few things:

Using the conda framework build (`pythonw`)

### v2.2.3:

If I run:

```
>>> import matplotlib
>>> matplotlib.use('macosx')
>>> import matplotlib.pyplot as plt
```

Similarly:

```
>>> from matplotlib.backends import _macosx

```

The terminal loses focus and the icon "bounces" if I use the conda framework build.  

Conclusion: the bouncing has always been there if you specify "macosx" and use the framework build. So @tacaswell's issue 1 above is not something we should consider fixing as a bug fix for 3.0.x.  If someone wants to figure out this behaviour, they should feel welcome to do so.

As @dstansby points out above the origin of this behaviour is in the `verify_framework()` routine in `_macosx.m`.  If you remove this check, then the icon is not activated until draw time.  

### v3.0.0

The issue is that with the new changes, import pyplot calls `_get_running_interactive_framework`, even if `mpl.use` has been called.   `_get_running_interactive_framework` tries to  `import _macosx` which checks that the framework build is being used, and pops up the icon to do so.  It needs to do this because the way to tell if a macosx GUI is being used is to call _macosx.event_loop_is_running().  

@anntzer or @tacaswell will have to explain the logic here, but I think the idea is that if a backend has not been set, we should check to see if Matplotlib is supposed to be running in GUI.  But it seems to be checking even if the backend *has* been set, and that is what is causing the problem here.  

I *don't* understand the new backend switching logic, but we either need to not call `_get_running_interactive_framework` or move `_macosx.m/verify_framework()` out of the init, so it doesn't get called when we `import _macosx` and call it somewhere else.  

Assigning the folks who understand this logic better, but more than willing to help test changes...

**Comment by anntzer** [comment]
I'll explain the current design and why it is as it is.

Before matplotlib 3.0, we shipped (i.e., Matplotlib wheels, conda packages, linux packages, etc.) a default matplotlibrc file with a `backend: ...` entry.  Depending on the build details, common values for that entry were be "agg" (not great for interactive use), "macosx" (only for OSX builds, doesn't work on non-framework builds), "tkagg" (works nearly everywhere although some linuxes distribute tkinter separately, and fails on headless systems), or "qt5agg" (done by conda but adds another (heavy!) dependency on PyQt).

Given the issues described above, for matplotlib 3.0, we changed the setup so that the default matplotlibrc does *not* include a `backend: ...` entry, and added the behavior below when this entry indeed doesn't exist.  Note that if the end user (or third-party packager) provides their own matplotlibrc with a `backend: ...` entry, nothing changes.

If there is no `backend: ...` entry, try to *import* backends in a statically defined order, namely macosx, qt5agg, ..., ending with agg -- a process we termed "backend fallback".  If a backend fails to load with an ImportError, move on to the next one.  If a backend requires a GUI framework that is incompatible with the currently running GUI framework (if any) (see below), also skip it.  The first non-skipped backend becomes the default backend.  The result is that when running on an OSX framework build, we get the macosx backend; otherwise, if PyQt5 is installed, we get qt5agg; etc.; if running headless, we get agg.  Ignoring implementation questions for now, this is indeed the desired behavior (see the numerous issues that previously existed regarding this topic).

How do we know whether "the backend requires a GUI framework that is incompatible with the currently running GUI framework (if any)"?  Backends have been changed to additionally expose a `required_framework` attribute, that is a string representing the framework ("macosx", "qt5", etc.) or None (for GUI-less backends); conversely, there is a `_get_running_interactive_framework` function that checks what the currently running framework is (if any); we just compare the result of `_get_running_interactive_framework()` to `backend_mod.required_framework`.

However, it is clear that this comparison must occur during backend fallback.  Therefore, moving the check down to Canvas or Manager instantiation time, as proposed by #12557, won't work, in the sense that backend fallback won't be able to detect some cases (namely, non-framework Pythons that don't have a `backend: ...` in their matplotlibrc) where the macosx backend shouldn't be loaded, and will load it incorrectly.

Really, the main necessity of the check is to know whether, yes or not, the current Python build is a framework Python build or not.  I'd hope this is something that can be done without triggering rocket bounces, but I don't know enough about OSX to tell.  It is somewhat sad, however, that the "official" way to perform that check (checking some stuff in the sysconfig module) doesn't work because conda's "framework" build is not a real framework build but something that fakes it :/ (

**Comment by defjaf** [comment]
An interesting point that may have not been noted before: 

Under the `python2` REPL
```
from matplotlib.backends import _macosx
```
does *not* bounce the icon --  the icon appears, but it doesn't bounce. But it does appear and bounce under `ipython2` or `jupyter` with a `python2` kernel.

**Comment by jklymak** [comment]
OK, but thats 2.3.3 behaviour.  Lets try to stick w/ the changes in 3.0.0 for this issue, which is that the icon gets activated even when it should not.  

If you think the old 2.2.3 behaviour is also undesirable, maybe open a new issue?

**Comment by jklymak** [comment]
OK, what chenged wrt to the rocket is that we took out the chunk of code that compiled if we were in 10.6 or greater in #11850.  Thats whats changed from 2.2.3 and is causing the bouncing behaviour.

**Comment by jklymak** [comment]
@defjaf Are you able to test master?  #12603 should fix the bouncing for most backends except "macosx", but it would be useful for more people to try.

**Comment by tacaswell** [comment]
This looks to be partially resolved (should only affect people who intend to use the macosx backend?) pushing to next milestone as we want to get 3.0.1 out the door.

**Comment by tacaswell** [comment]
I wonder if this is somehow related to readline.....

**Comment by anntzer** [comment]
> Alternatively the call to SetFrontProcess might be causing the bouncing. Maybe if there is a way to un-set the front process (set it back to what it was before that call?) that would stop the bouncing.

Indeed, perhaps it'll help to first call GetFrontProcess (https://developer.apple.com/documentation/applicationservices/1501050-getfrontprocess?language=objc) before doing all that manipulation, and restore it at the end?

**Comment by jklymak** [comment]
I'm going to close because I think this is as good as we are going to get for now.  There is a PR for 3.1 (#12633) that might make life even better for mac backend users...

**Comment by Zweedeend** [comment]
I had this problem in matplotlib==3.0, upgrading to 3.1 solved my problem.

**Comment by neilyoung** [comment]
3.4.2 bounce bounce bounce :/

**Comment by jklymak** [comment]
@neilyoung please open a new issue with version info, and exactly how you installed Matplotlib and python.  Thanks!

**Comment by neilyoung** [comment]
Thanks for the response. Meanwhile I found a work around by `plt.pause(0.0001)`. For the files: plotlib is 3.4.2, installed globally under python 3.9 at macOS 11.4

**Comment by jklymak** [comment]
How did you install python 3.9, and what does "installed globally" mean?

**Comment by neilyoung** [comment]
Sorry for not being specific enough. 3.9 comes with brew, install globally means "sudo python -m pip install matplotlib==3.4.2"

**Comment by jklymak** [comment]
I cannot reproduce with python3 3.9.1 from homebrew, neither by calling 

`python3 test.py`

```python
import matplotlib.pyplot as plt 
```

nor via the REPL.  I did not install globally, because that is not a good idea in general, so _maybe_ that is the difference? 

Again, please open a new issue, and explicitly state what you did to install python and what you did to install matplotlib.  Further state exactly what steps you took to get the bouncing doc.

**Comment by neilyoung** [comment]
This is not my app :)

I'm running a script in a "for" loop, which updates dots and lines and calls  "plt.draw()". This script (not mine) is supposed to work on other OS. 

Again. I'm through with this. For me the pause works.

Thanks. Don't worry
